keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+2. Only consider `properties` fields in the schema which actually successfully validate a given object (then `d` would be pruned in example 3).
+3. Only consider `properties` fields outside of `anyOf`, `allOf`, `oneOf`, `not` during pruning.
+4. Only consider `properties` fields outside of `anyOf`, `allOf`, `oneOf`, `not` during pruning, but enforce that every `properties` key inside `anyOf`, `allOf`, `oneOf`, `not` also appears outside all of those.
+
+From these options:
+1. leads to fields being kept from pruning although they only appear in branches of the OpenAPI validation schema which do not contribute to validation of an object.
+2. is ambiguous if you have multiple branches of `anyOf` validating. Should we drop the fields of the first or the second branch?
+3. leads to surprising pruning of fields that the user forgot to specify outside of propositional logic of `anyOf`, `allOf`, `oneOf`, `not`.
+4. forces the user to re-specify all those properties outside of `anyOf`, `allOf`, `oneOf`, `not` which appear inside of them. The outcome will match that of 1, with the difference that it makes the skeleton explicit.
+
+**Here we propose not to follow 2 and 3 due to ambiguity of 2 and the danger of user mistakes of 3. Both 1 and 4 lead to the same pruning behaviour and only differ in whether the user has to re-specify property keys or whether they are automatically derived.**
+
+### Restrictions on Defaults
+
+The OpenAPI validation schema allows to define defaults via the `default` field. For example:
+
+```json
+{"properties": {"a": {"default": {"x": 42}}}}
+```
+
+An empty object `{}` would be defaulted to `{"a":{"x":42}}`
+
+
+Obviously, the default values should validate. But OpenAPI validation schemata do not enforce that. Also one default in a branch of `anyOf`, `allOf`, `oneOf` might force another validation run onto another branch of these quantors.
+
+Even worse: forcing onto another branch might make more defaults applicable. In other words, we would need to run validation with defaulting until it converges. This is a hint that the semantics of defaulting in full OpenAPI schemata are not well defined either.
+
+**We propose to allow defaults only outside of `anyOf`, `allOf`, `oneOf`, `not`.**
+
+Without that, we would get ambiguous defaulting semantics:
+
+#### Example 4:
+
+> Assume 
+>```json
+>{
+>    "anyOf":[
+>       {"default": 1},
+>       {"default": 2}
+>    ]
+>}
+>```
+>
+> Which branch should be chosen while applying the defaults.
+
+#### Example 5:
+
+> Assume
+>
+>```json
+>{
+>  "anyOf":[
+>      {"properties": {"a": {"pattern": "1"}, "b": {"type": "string"}, "d": {"default": 1}}},
+>      {"properties": {"a": {"pattern": "2"}, "c": {"type": "string"}, "d": {"default": 2}}}
+>  ]
+>}
+>```
+>
+> Take an object `{"a": "1", "b":"foo", "c":"bar"}`. Assuming we follow pruning option 1, it is pruned to `{"a": "1", "b":"foo", "c":"bar"}` and to default to `{"a":"1", "b": "foo", "c":"bar", "d": 1}`.
```

Comment:
**David Eads:** I would not have expected to be pruned. I'd expect `c` to be kept and `d=1` (first rule in what I described).

I'm also ok with restricting this for a first impl.

**Antoine Pelisse:** how would this work with `additionalProperties: false`?

**Stefan Schimanski:** We forbid that additionProperties and properties are used in parallel for the same object/map. So `additionalProperties: false` means to prune everything.

Reply:
@deads2k added this as a comment further down that this is the starting point with the options for relaxing this later if needed.

Reply:
We need to engineer our schema-acceptance criteria so we never face this question, I think.